File: 2022/day8/day8.js
const fs = require('fs');
const path = require('path');

let DIRS = [
  { y: -1, x: 0 },
  { y: 1, x: 0 },
  { y: 0, x: -1 },
  { y: 0, x: 1 },
];

function isInside(grid, x, y) {
  return x >= 0 && y >= 0 && x < grid[0].length && y < grid.length;
}

function collectDirsTrees(grid, x, y) {
  let positions = [[], [], [], []];
  for (let i = 0; i < DIRS.length; i++) {
    let dir = DIRS[i];
    let [x2, y2] = [x, y];
    while (true) {
      x2 += dir.x;
      y2 += dir.y;
      if (isInside(grid, x2, y2)) positions[i].push(grid[y2][x2]);
      else break;
    }
  }
  return positions;
}

function part1(input) {
  /** @type string[] */
  let grid = input.trim().split('\n');
  let visible = 0;
  let decreasingFrom = from => list => list.every(num => num < from);

  for (let y = 0; y < grid.length; y++) {
    let row = grid[y];
    for (let x = 0; x < row.length; x++) {
      let dirsTrees = collectDirsTrees(grid, x, y);
      let isVisible = dirsTrees.some(decreasingFrom(grid[y][x]));
      if (isVisible) {
        visible++;
      }
    }
  }
  return visible;
}

function part2(input) {
  /** @type string[] */
  let grid = input.trim().split('\n');
  let allScores = [];
  let distance = from => list => {
    let index = list.findIndex(tree => tree >= from);
    return index === -1 ? list.length : index + 1;
  };

  for (let y = 0; y < grid.length; y++) {
    let row = grid[y];
    for (let x = 0; x < row.length; x++) {
      let dirsTrees = collectDirsTrees(grid, x, y);
      let scenicScore = dirsTrees.map(distance(grid[y][x])).reduce((a, b) => a * b);
      allScores.push(scenicScore);
    }
  }
  return Math.max(...allScores);
}

let input = fs.readFileSync(path.resolve(__dirname, './input.txt'), 'utf8').replaceAll('\r', '');

// input = `
// 30373
// 25512
// 65332
// 33549
// 35390`;

console.log('part1:', part1(input));
console.log('part2:', part2(input));

File: 2022/day8/day8.clj
(require '[clojure.string :as str])

(def input-raw (slurp (str "2022/" (-> *file* (str/split #"/") butlast last) "/input.txt")))
#_(def input-raw "30373
25512
65332
33549
35390")

(def input
  (->> input-raw
       str/trim
       str/split-lines
       (map #(map read-string (str/split % #"")))))

(defn find-index [pred coll]
  (first (keep-indexed #(when (pred %2) %1) coll)))

(def dirs [{:y -1 :x 0}
           {:y 1 :x 0}
           {:y 0 :x -1}
           {:y 0 :x 1}])

(defn is-inside [grid coord]
  (every? true? ((juxt #(>= (:x %) 0)
                       #(>= (:y %) 0)
                       #(< (:y %) (count grid))
                       #(< (:x %) (count grid))) coord)))

(defn collect-dirs [grid coord]
  (map (fn [dir]
         (->> coord
              (iterate (partial merge-with + dir))
              (drop 1)
              (take-while (partial is-inside grid))))
       dirs))

(defn decreasing-from [from]
  (fn [xs] (every? (partial > from) xs)))

(defn distance [from]
  (fn [xs]
    (let [found (find-index (partial <= from) xs)]
      (if (nil? found)
        (count xs)
        (inc found)))))

(defn get-value [grid coord]
  (nth (nth grid (:y coord)) (:x coord)))

(defn part1 [input]
  (->> input
       (map-indexed
        (fn [y row]
          (map-indexed
           (fn [x n] (if (some (decreasing-from n)
                               (map (fn [dirPath]
                                      (map (partial get-value input) dirPath))
                                    (collect-dirs input {:x x :y y})))
                       1
                       0))
           row)))
       flatten
       (apply +)))

(defn part2 [input]
  (->> input
       (map-indexed
        (fn [y row]
          (map-indexed
           (fn [x n] (->> (map (distance n) (map (fn [dirPath] (map (partial get-value input) dirPath))
                                                 (collect-dirs input {:x x :y y})))
                          (apply *)))
           row)))
       flatten
       (apply max)))

(prn "part1:" (part1 input))
(prn "part2:" (part2 input))
